src/layers/ipv4.rs

```diff
@@ -305,6 +321,42 @@ impl Layer for IPv4 {
     fn short_name(&self) -> &'static str {
         "ip"
     }
+
+    fn stack_and_encode(&mut self, next_layer: &[u8], info: &str) -> Result<Vec<u8>, Error> {
+        self.len = self.hdr_len as u16 + next_layer.len() as u16;
+        self.proto = match info {
+            "raw" => IPPROTO_RAW,
+            _ => self.proto,
+        };
+
+        let mut result = vec![];
```

Comment:
Reserve - the header bytes when creating the `Vec` - this should then avoid alloc and moves during extend and then a simple `mem::replace` at offsets. 

Worth benchmarking both approaches though. Let this be as it is (we can do some feature compilation during initial dev to see which approach is better)